libindy/src/services/pool/catchup.rs

```diff
@@ -115,40 +118,67 @@ impl CatchupHandler {
-                let &(ref target_mt_root, target_mt_size) = most_popular_vote;
-                let cur_mt_size = self.merkle_tree.count();
-                let cur_mt_hash = self.merkle_tree.root_hash().to_base58();
-                if target_mt_size == cur_mt_size {
-                    if cur_mt_hash.eq(target_mt_root) {
-                        return Ok(CatchupProgress::NotNeeded);
-                    } else {
-                        return Err(PoolError::CommonError(CommonError::InvalidState(
-                            "Ledger merkle tree doesn't acceptable for current tree.".to_string())));
+                return match self._try_to_catch_up(most_popular_vote) {
+                    //WARNING: If we receive invalid cache then it means that cache was cleared. We need to rebuild merkle tree and try to catch up once again
+                    Err(PoolError::InvalidCacheCleared) => {
+                        self.merkle_tree = PoolWorker::restore_merkle_tree_from_pool_name(&self.pool_name)?;
+                        self._try_to_catch_up(most_popular_vote)
                     }
-                } else if target_mt_size > cur_mt_size {
-                    self.target_mt_size = target_mt_size;
-                    self.target_mt_root = target_mt_root.from_base58().map_err(|_|
-                        CommonError::InvalidStructure(
-                            "Can't parse target MerkleTree hash from nodes responses".to_string()))?;
-                    return Ok(CatchupProgress::ShouldBeStarted);
-                } else {
-                    return Err(PoolError::CommonError(CommonError::InvalidState(
-                        "Local merkle tree greater than mt from ledger".to_string())));
+                    a => a
                 }
             }
         }
         Ok(CatchupProgress::InProgress)
     }
 
+    fn _try_to_catch_up(&mut self, most_popular_vote: &(String, usize, Option<Vec<String>>)) -> Result<CatchupProgress, PoolError> {
+        let &(ref target_mt_root, target_mt_size, ref hashes) = most_popular_vote;
+        let cur_mt_size = self.merkle_tree.count();
+        let cur_mt_hash = self.merkle_tree.root_hash().to_base58();
+        if target_mt_size == cur_mt_size {
+            if cur_mt_hash.eq(target_mt_root) {
+                return Ok(CatchupProgress::NotNeeded);
+            } else {
+                PoolWorker::drop_saved_txns(&self.pool_name)?;
+                return Err(PoolError::CommonError(CommonError::InvalidState(
+                    "Ledger merkle tree doesn't acceptable for current tree.".to_string())));
+            }
+        } else if target_mt_size > cur_mt_size {
+            self.target_mt_size = target_mt_size;
+            self.target_mt_root = target_mt_root.from_base58().map_err(|_|
+                CommonError::InvalidStructure(
+                    "Can't parse target MerkleTree hash from nodes responses".to_string()))?;
+            match hashes {
+                &None => (),
```

Comment:
missed consistency proof is something like protocol error. My personal vote to throw error for whole catchup process here
@vimmerru do you have another vision?